### PR TITLE
Add UniqueFactory for creating random string or int from given range

### PR DESCRIPTION
**Pull request by mleczakm**
<h2>Change Log</h2>
<div id="change-log">
  <h4>Added</h4>
  <ul id="added">
     <li>UniqueFactory as source of random ints/strings with given length</li> 
  </ul> 
  <h4>Fixed</h4>  
  <ul id="fixed">
    
  </ul>
  <h4>Changed</h4>
  <ul id="changed">
    
  </ul>  
  <h4>Removed</h4>
  <ul id="removed">
    
  </ul>
  <h4>Deprecated</h4>
  <ul id="deprecated">
    
  </ul>  
  <h4>Security</h4> 
  <ul id="security">
    
  </ul>     
</div>
<hr/>

<h2>Description</h2>

Solves https://github.com/flow-php/flow/issues/1124

### CODE REVIEW

File: src/core/etl/src/Flow/ETL/UniqueFactory.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL;
+
+class UniqueFactory
```
Comment (stloyd) [inline src/core/etl/src/Flow/ETL/UniqueFactory.php]:
```suggestion
final class UniqueFactory
```

File: src/core/etl/src/Flow/ETL/UniqueFactory.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+{
+    public static function int(int $min, int $max) : int
```
Comment (stloyd) [inline src/core/etl/src/Flow/ETL/UniqueFactory.php]: Maybe also:
```suggestion
    private function __construct()
    {}

    public static function int(int $min, int $max) : int
```
Cause this class has only static methods.

File: src/core/etl/src/Flow/ETL/UniqueFactory.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        return \random_int($min, $max);
+    }
+
+    /** @param int<1, max> $int */
+    public static function string(int $int) : string
+    {
+        $bytes = (int) \ceil($int / 2);
+        $bytes >= 1 ?: $bytes = 1;
+
+        return \mb_substr(\bin2hex(\random_bytes($bytes)), 0, \max(0, $int));
```
Comment (stloyd) [inline src/core/etl/src/Flow/ETL/UniqueFactory.php]: I think we don't need multibyte here, do we?

File: src/core/etl/tests/Flow/ETL/Tests/Integration/DataFrame/PartitioningTest.php [hunk 4]
```diff
@@ -163,13 +163,13 @@ public function test_partition_by_partitions_order() : void
                     function (int $i) : array {
                         $data = [];
 
-                        $maxItems = \random_int(2, 10);
```
Comment (stloyd) [inline src/core/etl/tests/Flow/ETL/Tests/Integration/DataFrame/PartitioningTest.php:166]: More I look at this, more I think I would add new function for unique int/string to "hide" object call & improve DX perspective, WDYT?

**Comment by norberttech** [comment]
The PR look good and we could merge it like that (beside few small things @stloyd pointed out) but I think we might be able to still improve even more. Here are my thoughts. 

**Naming** - now when I look at it something is missing in the name, like `RandomValueGenerator` would be more descriptive. 

**Dependency Injection** - I think I would turn `RandomValueGenerator` into an interface with `NativePHPRandomValueGenerator` that we can simply inject into objects that would need a unique values (I'm thinking about caching strategies here and more advanced data processing algorithms). 

**Scalar Functions** - Additionally (can go to a standalone PR) we could also create Flow scalar functions that could be used during data processing, like: 

```php
df()
   ->read(...) 
   ->withEntry('random_string', random_string(length: 16))
   ->write(to_output())
   ->run()

df()
   ->read(...) 
   ->withEntry('random_string', random_string(length: ref('some_int_column'))))
   ->write(to_output())
   ->run()
```

This function would implement `Flow\ETL\Function\ScalarFunction` and could take `$lenght` as another `ScalarFunction` so this way length could be taken from a column or output of chained functions (some calculations). Additionally it would also need `RandomValueGenerator $generator = new NatievPHPRandomValueGenerator()` in case anyone would like to change the generation strategy. 

**DSL Functions** - as pointed by @stloyd we probably should also add `generate_random_string(int $length = 16)` and `generate_random_int(int $start = PHP_INT_MIN, int $end = PHP_INT_MAX, RandomValueGenerator $generator = new NatievPHPRandomValueGenerator())` functions so we can use them in tests instead of static factories. 

Thoughts?

File: src/core/etl/src/Flow/ETL/UniqueFactory.php (hunk 3 above)
Comment (mleczakm) [inline src/core/etl/src/Flow/ETL/UniqueFactory.php]:
```suggestion
        return \substr(\bin2hex(\random_bytes($bytes)), 0, \max(0, $int));
```

**Review by norberttech — CHANGES_REQUESTED** [review]
Some changes described above are required to merge this PR :)

**Review by norberttech — APPROVED** [review]

**Pull request merged**